### PR TITLE
Pass the Authorization header across to the redirected request

### LINKED ISSUE #2711: GitHub Access Token not being used causing Rate Limit to be hit
* carthage install method: Pre-installed on Azure Pipelines Hosted macOS agent
* `which carthage`: 
* `carthage version`: 0.29.0
* `xcodebuild -version`: Xcode 10.1 Build version 10B61
* Are you using `--no-build`?  No
* Are you using `--no-use-binaries`? No
* Are you using `--use-submodules`? No
* Are you using `--cache-builds`? No
* Are you using `--new-resolver`? No

**Cartfile**
```
github "aws/aws-sdk-ios" == 2.6.31
github "daltoniam/Starscream" ~> 3.0
github "socketio/socket.io-client-swift" == 13.3.1
github "kishikawakatsumi/KeychainAccess" == 3.1.2
```

**Carthage Output**
```
2019-02-19T08:21:37.4263280Z *** No Cartfile.resolved found, updating dependencies
2019-02-19T08:21:37.4628150Z *** Fetching Starscream
2019-02-19T08:21:38.4336330Z *** Fetching socket.io-client-swift
2019-02-19T08:21:39.0465680Z *** Fetching KeychainAccess
2019-02-19T08:21:39.6142120Z *** Cloning aws-sdk-ios
2019-02-19T08:21:52.5121650Z *** Checking out aws-sdk-ios at "2.6.31"
2019-02-19T08:21:52.5157120Z *** Checking out Starscream at "3.0.6"
2019-02-19T08:21:52.5158360Z *** Checking out socket.io-client-swift at "v13.3.1"
2019-02-19T08:21:52.5158740Z *** Checking out KeychainAccess at "v3.1.2"
2019-02-19T08:21:53.1632780Z *** xcodebuild output can be found in /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/carthage-xcodebuild.B9zfOW.log
2019-02-19T08:21:53.6859200Z *** Skipped downloading aws-sdk-ios.framework binary due to the error:
2019-02-19T08:21:53.6860170Z 	"API rate limit exceeded for A.B.C.D. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)"
```

**Actual outcome**
Carthage did not successfully download the frameworks from GitHub as it appears to have ignored the supplied GITHUB_ACCESS_TOKEN.

**Expected outcome**
Carthage should use the access token to make use of higher rate limits.

### More Info
For some reason it looks like the access token is not being used, I have a script running prior to `carthage bootstrap` to output the current rate limits (both unauthenticated and authenticated, using the same access token environment variable as Carthage is looking for):

**Unauthenticated:**
```
2019-02-19T08:21:35.7419280Z {
2019-02-19T08:21:35.7419740Z   "resources": {
2019-02-19T08:21:35.7420170Z     "core": {
2019-02-19T08:21:35.7420350Z       "limit": 60,
2019-02-19T08:21:35.7420570Z       "remaining": 0,
2019-02-19T08:21:35.7420630Z       "reset": 1550568095
2019-02-19T08:21:35.7422110Z     },
2019-02-19T08:21:35.7422560Z     "search": {
2019-02-19T08:21:35.7422890Z       "limit": 10,
2019-02-19T08:21:35.7423020Z       "remaining": 10,
2019-02-19T08:21:35.7423150Z       "reset": 1550564555
2019-02-19T08:21:35.7424870Z     },
2019-02-19T08:21:35.7425560Z     "graphql": {
2019-02-19T08:21:35.7426230Z       "limit": 0,
2019-02-19T08:21:35.7426410Z       "remaining": 0,
2019-02-19T08:21:35.7426820Z       "reset": 1550568095
2019-02-19T08:21:35.7427100Z     },
2019-02-19T08:21:35.7427420Z     "integration_manifest": {
2019-02-19T08:21:35.7427760Z       "limit": 5000,
2019-02-19T08:21:35.7427890Z       "remaining": 5000,
2019-02-19T08:21:35.7428030Z       "reset": 1550568095
2019-02-19T08:21:35.7428930Z     }
2019-02-19T08:21:35.7429090Z   },
2019-02-19T08:21:35.7429170Z   "rate": {
2019-02-19T08:21:35.7429400Z     "limit": 60,
2019-02-19T08:21:35.7429490Z     "remaining": 0,
2019-02-19T08:21:35.7429590Z     "reset": 1550568095
2019-02-19T08:21:35.7429770Z   }
2019-02-19T08:21:35.7429860Z }
```

**Authenticated:**
```
2019-02-19T08:21:36.3992550Z {
2019-02-19T08:21:36.3992950Z   "resources": {
2019-02-19T08:21:36.3993390Z     "core": {
2019-02-19T08:21:36.3993710Z       "limit": 5000,
2019-02-19T08:21:36.3994040Z       "remaining": 4952,
2019-02-19T08:21:36.3994460Z       "reset": 1550566157
2019-02-19T08:21:36.3994780Z     },
2019-02-19T08:21:36.3995170Z     "search": {
2019-02-19T08:21:36.3995590Z       "limit": 30,
2019-02-19T08:21:36.3996190Z       "remaining": 30,
2019-02-19T08:21:36.3997830Z       "reset": 1550564556
2019-02-19T08:21:36.3998200Z     },
2019-02-19T08:21:36.3998560Z     "graphql": {
2019-02-19T08:21:36.3998940Z       "limit": 5000,
2019-02-19T08:21:36.3999020Z       "remaining": 4994,
2019-02-19T08:21:36.3999170Z       "reset": 1550568046
2019-02-19T08:21:36.3999270Z     },
2019-02-19T08:21:36.3999940Z     "integration_manifest": {
2019-02-19T08:21:36.4000140Z       "limit": 5000,
2019-02-19T08:21:36.4000250Z       "remaining": 5000,
2019-02-19T08:21:36.4000330Z       "reset": 1550568096
2019-02-19T08:21:36.4000470Z     }
2019-02-19T08:21:36.4000550Z   },
2019-02-19T08:21:36.4000680Z   "rate": {
2019-02-19T08:21:36.4000780Z     "limit": 5000,
2019-02-19T08:21:36.4000850Z     "remaining": 4952,
2019-02-19T08:21:36.4000990Z     "reset": 1550566157
2019-02-19T08:21:36.4001080Z   }
2019-02-19T08:21:36.4001150Z }
```

Then I run the command:
`carthage bootstrap --platform iOS`

And then query the rate limits again:

**Unauthenticated:**
```
2019-02-19T08:22:04.5807100Z {
2019-02-19T08:22:04.5808420Z   "resources": {
2019-02-19T08:22:04.5808780Z     "core": {
2019-02-19T08:22:04.5809160Z       "limit": 60,
2019-02-19T08:22:04.5809420Z       "remaining": 0,
2019-02-19T08:22:04.5809730Z       "reset": 1550568124
2019-02-19T08:22:04.5810050Z     },
2019-02-19T08:22:04.5810270Z     "search": {
2019-02-19T08:22:04.5810680Z       "limit": 10,
2019-02-19T08:22:04.5810930Z       "remaining": 10,
2019-02-19T08:22:04.5811260Z       "reset": 1550564584
2019-02-19T08:22:04.5811530Z     },
2019-02-19T08:22:04.5812040Z     "graphql": {
2019-02-19T08:22:04.5812160Z       "limit": 0,
2019-02-19T08:22:04.5812510Z       "remaining": 0,
2019-02-19T08:22:04.5812760Z       "reset": 1550568124
2019-02-19T08:22:04.5813060Z     },
2019-02-19T08:22:04.5813570Z     "integration_manifest": {
2019-02-19T08:22:04.5816880Z       "limit": 5000,
2019-02-19T08:22:04.5825830Z       "remaining": 5000,
2019-02-19T08:22:04.5826410Z       "reset": 1550568124
2019-02-19T08:22:04.5826910Z     }
2019-02-19T08:22:04.5827470Z   },
2019-02-19T08:22:04.5827970Z   "rate": {
2019-02-19T08:22:04.5828440Z     "limit": 60,
2019-02-19T08:22:04.5829020Z     "remaining": 0,
2019-02-19T08:22:04.5829500Z     "reset": 1550568124
2019-02-19T08:22:04.5829980Z   }
2019-02-19T08:22:04.5830570Z }
```

**Authenticated:**
```
2019-02-19T08:22:05.1090500Z {
2019-02-19T08:22:05.1090970Z   "resources": {
2019-02-19T08:22:05.1091380Z     "core": {
2019-02-19T08:22:05.1091880Z       "limit": 5000,
2019-02-19T08:22:05.1092290Z       "remaining": 4948,
2019-02-19T08:22:05.1092690Z       "reset": 1550566157
2019-02-19T08:22:05.1093210Z     },
2019-02-19T08:22:05.1093610Z     "search": {
2019-02-19T08:22:05.1094080Z       "limit": 30,
2019-02-19T08:22:05.1094520Z       "remaining": 30,
2019-02-19T08:22:05.1094920Z       "reset": 1550564585
2019-02-19T08:22:05.1095510Z     },
2019-02-19T08:22:05.1095950Z     "graphql": {
2019-02-19T08:22:05.1096360Z       "limit": 5000,
2019-02-19T08:22:05.1096860Z       "remaining": 4994,
2019-02-19T08:22:05.1097270Z       "reset": 1550568046
2019-02-19T08:22:05.1097750Z     },
2019-02-19T08:22:05.1098640Z     "integration_manifest": {
2019-02-19T08:22:05.1099050Z       "limit": 5000,
2019-02-19T08:22:05.1099550Z       "remaining": 5000,
2019-02-19T08:22:05.1099970Z       "reset": 1550568125
2019-02-19T08:22:05.1100380Z     }
2019-02-19T08:22:05.1100880Z   },
2019-02-19T08:22:05.1101280Z   "rate": {
2019-02-19T08:22:05.1101670Z     "limit": 5000,
2019-02-19T08:22:05.1102190Z     "remaining": 4948,
2019-02-19T08:22:05.1102590Z     "reset": 1550566157
2019-02-19T08:22:05.1103060Z   }
2019-02-19T08:22:05.1103480Z }
```

As shown the rate limit for authenticated requests is still well above 0 so it shouldn't have an issue downloading the frameworks.

### PR DESCRIPTION
**Pull request by NxSoftware**
Should resolve #2711

### CODE REVIEW

File: Source/CarthageKit/GitHub.swift [hunk 1]
```diff
@@ -167,6 +167,18 @@ extension URLSession {
...
+	}
+}
+
+class GitHubURLSessionDelegate: NSObject, URLSessionTaskDelegate {
+	func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
+		let originalRequest = task.originalRequest
+		var newRequest = request
+		if let auth = originalRequest?.value(forHTTPHeaderField: "Authorization") {
+			newRequest.setValue(auth, forHTTPHeaderField: "Authorization")
+		}
```
Comment (mdiep) [inline Source/CarthageKit/GitHub.swift]: Can you add a check to only do this if you're redirected to the same domain?

I don't think it should matter, but I'd like to be conservative with peoples' credentials.

File: Source/CarthageKit/GitHub.swift [hunk 2]
```diff
@@ -167,6 +167,18 @@ extension URLSession {
 		let configuration = URLSessionConfiguration.default
 		configuration.connectionProxyDictionary = Proxy.default.connectionProxyDictionary
 
-		return URLSession(configuration: configuration)
+		let delegate = GitHubURLSessionDelegate()
+		return URLSession(configuration: configuration, delegate: delegate, delegateQueue: OperationQueue.current)
+	}
+}
+
+class GitHubURLSessionDelegate: NSObject, URLSessionTaskDelegate {
```
Comment (mdiep) [inline Source/CarthageKit/GitHub.swift]: Can you make this `private final`?

**Comment by mdiep** [comment]
I'm glad you were able to track this down! A couple small tweaks, but then we can merge this. 👍

File: Source/CarthageKit/GitHub.swift (hunk 1 above)
Comment (NxSoftware) [reply Source/CarthageKit/GitHub.swift]: Yeh good idea. I also added a couple of tests around this logic.

File: Source/CarthageKit/GitHub.swift (hunk 2 above)
Comment (NxSoftware) [reply Source/CarthageKit/GitHub.swift]: Made it `internal final` so it'll remain accessible to tests.

**Comment by mdiep** [comment]
Awesome! Thanks for the contribution. ✨

**Pull request merged**